File: Prog 3 - Trabalho 1 (Java)/src/trabalho1/EleicaoDeVereadores.java
package trabalho1;

import java.util.*;

/**
 * Essa classe serve para armazenar e trabalhar com as informações adquiridas com a classe "Leitor"
 * 
 */

public class EleicaoDeVereadores {
	
	private LinkedList<Vereador> listaTodosOsVereadores = new LinkedList<Vereador>();
	private LinkedList<Coligacao> listaTodasAsColigacoes = new LinkedList<Coligacao>();
	private LinkedList<Partido> listaTodosOsPartidos = new LinkedList<Partido>();
	
	/**
	 * Esse método cria um novo vereador a cada linha lida do arquivo de entrada, podendo ou não criar novo partido e coligação
	 * depedendo da necessidade.
	 * <P>
	 * É o método principal dessa classe, pois ela chamará todos os outros quando necessário, com exceção de "adicionaColigacoesDeUmPartido".
	 * @param leitor 
	 */
	
	public void novoVereador(Leitor leitor){
		Vereador novoVereador = new Vereador();
		novoVereador.setId(leitor.getId());
		novoVereador.setNome(leitor.getNomeVereador());
		novoVereador.setNum(leitor.getNumero());	
		novoVereador.setVotos(leitor.getVotos());
		novoVereador.setEleito(leitor.isEleito());												
		Partido partido = verificaPartido(leitor);			//Chama método para chamar e criar (se necessário) um partido, para assim
		novoVereador.setPartido(partido);					//o vereador ser adicionado a lista do partido, e o partido adicionado ao vereador
		if (!(leitor.getColigacao()==null)){
			novoVereador.setColigacao(partido.getColigacao());		//Caso o vereador não tenha coligação, fica como null
		}
		partido.getVereadoresDoPartido().add(novoVereador);					
		listaTodosOsVereadores.add(novoVereador);
	}
	
	public LinkedList<Vereador> getTodosOsVereadores() {
		return listaTodosOsVereadores;
	}
	
	public LinkedList<Partido> getPartidos() {
		return listaTodosOsPartidos;
	}
	
	/**
	 * Método para criação de partido, chamado quando não encontra um partido de mesmo nome na lista.
	 * @param leitor 
	 * @return Tipo partido, 
	 */
	
	public Partido criaPartido (Leitor leitor){
		Partido novoPartido = new Partido();
		novoPartido.setNome(leitor.getPartido());
		novoPartido.setVotosTotal(leitor.getVotos());
		if (leitor.isEleito()){
			novoPartido.setVereadoresEleitos(1);
		}
		return novoPartido;
	}
	
	/**
	 * Verifica a existência do partido na lista e se o mesmo possui coligação.
	 * Caso o partido não possua coligação, a variável 'Coligacao' fica como 'null'.
	 * <p>
	 * Caso encontre o partido, e esse partido possui uma coligação, é criado
	 * uma váriavel tipo 'Coligacao' que não é usada, porém necessária para chamar
	 * o método que irá modificar, se necessário, a quantidade de votos e vereadores
	 * eleitos
	 * @param leitor
	 * @return Partido
	 */
	
	public Partido verificaPartido (Leitor leitor){
		boolean existePartido=false;
		Partido novoPartido = new Partido();
		for (Partido partido : listaTodosOsPartidos){
			if (Objects.equals(leitor.getPartido(), partido.getNome())){
				existePartido=true;
				partido.setVotosTotal(partido.getVotosTotal()+leitor.getVotos());
				if (leitor.isEleito()){
					partido.setVereadoresEleitos(partido.getVereadoresEleitos()+1);
				}
				if (!(leitor.getColigacao()==null)){
					Coligacao coligacao = verificaColigacao(leitor);
				}												
				return partido;										
			}
		}
		if (!(existePartido)){
			novoPartido = criaPartido(leitor);
			listaTodosOsPartidos.add(novoPartido);
			if (!(leitor.getColigacao()==null)){
				Coligacao coligacao = new Coligacao();
				coligacao = verificaColigacao(leitor);
				coligacao.getPartidosDaColigacao().add(novoPartido);	
				novoPartido.setColigacao(coligacao);
			}
		}
		return novoPartido;
	}
	
	public LinkedList<Coligacao> getColigacoes(){
		return listaTodasAsColigacoes;
	}
	
	/**
	 * Método para criar novas coligações
	 * @param leitor
	 * @return Nova 'Coligacao' criada
	 */
	public Coligacao criaColigacao (Leitor leitor){
		Coligacao novaColigacao = new Coligacao();
		novaColigacao.setNome(leitor.getColigacao());
		novaColigacao.setVotosColigacao(leitor.getVotos());
		novaColigacao.getPartidosDaColigacao().add(criaPartido(leitor));
		if (leitor.isEleito()){
			novaColigacao.setVereadoresEleitosCol(1);
		}
		return novaColigacao;
	}
	
	/**
	 * Verifica se a coligação já foi criada e adicionada a lista, se tiver sido, é modificado as váriaveis
	 * de votos e vereadores eleitos. Caso contrário é criado uma nova coligação.
	 * 
	 * @param leitor
	 * @return tipo 'Coligacao'
	 */
	public Coligacao verificaColigacao (Leitor leitor){
		boolean existeColigacao = false;
		Coligacao novaColigacao = new Coligacao();
		for (Coligacao coligacao : listaTodasAsColigacoes){
			if (Objects.equals(leitor.getColigacao(), coligacao.getNome())){
				existeColigacao = true;
				coligacao.setVotosColigacao(coligacao.getVotosColigacao()+leitor.getVotos());
				if (leitor.isEleito()){
					coligacao.setVereadoresEleitosCol(coligacao.getVereadoresEleitosCol()+1);
				}
				return coligacao;
			}
		}
		if (!(existeColigacao)){
			novaColigacao = criaColigacao(leitor);
			listaTodasAsColigacoes.add(novaColigacao);
		}
		return novaColigacao;
	}
	
	/**
	 * Esse método foi criado apenas para adicionar os partidos que não possui coligação a lista de coligações.
	 */
	
	public void adicionaColigacoesDeUmPartido (){
		for (Partido partido : listaTodosOsPartidos){
			if (partido.getColigacao()==null){
				Coligacao coligacao = new Coligacao();
				coligacao.setNome (partido.getNome());
				coligacao.getPartidosDaColigacao().add(partido);
				coligacao.setVereadoresEleitosCol(partido.getVereadoresEleitos());
				coligacao.setVotosColigacao(partido.getVotosTotal());
				listaTodasAsColigacoes.add(coligacao);
			}
		}
	}
}

File: Prog 3 - Trabalho 2 (C++)/EleicaoDeVereadores.cpp

#include <list>
#include <string>
#include <iostream>
#include "EleicaoDeVereadores.h"
#include "Leitor.h"
#include "Partido.h"
#include "Vereador.h"
#include "Coligacao.h"

using namespace std;

void EleicaoDeVereadores::novoVereador (Leitor *leitor){
    Vereador *novoVereador = new Vereador;
    novoVereador->setId(leitor->getId());
    novoVereador->setNome(leitor->getNomeVereador());
    novoVereador->setNum(leitor->getNumero());
    novoVereador->setVotos(leitor->getVotos());
    novoVereador->setEleitos(leitor->isEleito());
    Partido *partido = verificaPartido(leitor);
    novoVereador->setPartido(partido);
    novoVereador->setColigacao(partido->getColigacao());
    partido->addNovoVereador(novoVereador);					
    listaTodosOsVereadores.push_back(novoVereador);
    if (leitor->isEleito()){
        eleitos+=1;
    }
}

Partido* EleicaoDeVereadores::criaPartido (Leitor* leitor){
    Partido *novoPartido = new Partido;
    novoPartido->setNome(leitor->getPartido());
    novoPartido->setVotosPartido(leitor->getVotos());
    if (leitor->isEleito()){
        novoPartido->setVereadoresEleitos(1);
    }
    return novoPartido;
}

Partido* EleicaoDeVereadores::verificaPartido (Leitor *leitor){
    bool existePartido=false;
    Partido *novoPartido = new Partido();
    for (list<Partido*>::iterator partido = listaTodosOsPartidos.begin(); partido != listaTodosOsPartidos.end(); partido++){
        Partido *aux = *partido;
        if (leitor->getPartido().compare(aux->getNome()) == 0){
            existePartido=true;
            aux->setVotosPartido(aux->getVotosPartido()+leitor->getVotos());
            if (leitor->isEleito()){
                aux->setVereadoresEleitos(aux->getVereadoresEleitos()+1);
            }
            Coligacao *coligacao = verificaColigacao(leitor);
            *partido = aux;
            return *partido;										
        }
    }
    if (!(existePartido)){
        novoPartido = criaPartido(leitor);
        listaTodosOsPartidos.push_back(novoPartido);
        Coligacao *coligacao = new Coligacao;
        coligacao = verificaColigacao(leitor);
        coligacao->addNovoPartido(novoPartido);	
        novoPartido->setColigacao(coligacao);
    }
    return novoPartido;
}

Coligacao* EleicaoDeVereadores::criaColigacao (Leitor* leitor){
    Coligacao* novaColigacao = new Coligacao;
    if (!leitor->getColigacao().compare("")==0){
        novaColigacao->setNome(leitor->getColigacao());
    }
    else{
        novaColigacao->setNome(leitor->getPartido());
    }
    novaColigacao->setVotosColigacao(leitor->getVotos());
    novaColigacao->addNovoPartido(criaPartido(leitor));
    if (leitor->isEleito()){
        novaColigacao->setVereadoresEleitos(1);
    }
    return novaColigacao;
}

Coligacao* EleicaoDeVereadores::verificaColigacao (Leitor* leitor){
    bool existeColigacao = false;
    Coligacao* novaColigacao = new Coligacao;
    for (list <Coligacao*>::iterator coligacao = listaTodasAsColigacoes.begin(); coligacao!=listaTodasAsColigacoes.end(); coligacao++){
        Coligacao *aux = *coligacao;
        if (!leitor->getColigacao().compare("")==0){
            if (leitor->getColigacao().compare(aux->getNome())==0){
                existeColigacao = true;
                aux->setVotosColigacao(aux->getVotosColigacao()+leitor->getVotos());
                if (leitor->isEleito()){
                    aux->setVereadoresEleitos(aux->getVereadoresEleitos()+1);
                }
                *coligacao = aux;
                return *coligacao;
            }
        }
        else {
            if (leitor->getPartido().compare(aux->getNome())==0){
                existeColigacao = true;
                aux->setVotosColigacao(aux->getVotosColigacao()+leitor->getVotos());
                if (leitor->isEleito()){
                    aux->setVereadoresEleitos(aux->getVereadoresEleitos()+1);
                }
                *coligacao = aux;
                return *coligacao;
            }
        }
    }
    if (!(existeColigacao)){
        novaColigacao = criaColigacao(leitor);
        listaTodasAsColigacoes.push_back(novaColigacao);
    }
    return novaColigacao;    
}

void EleicaoDeVereadores::ordenaVereadoresPorVotos(){
    Vereador *vereadorMaisVotado = new Vereador();
    list<Vereador*> vereadoresOrdenados;
    list<Vereador*>::iterator it;
    bool listaIncompleta = true;
    while (listaIncompleta){												
        listaIncompleta = false;										
        for (list<Vereador*>::iterator vereador = listaTodosOsVereadores.begin(); vereador != listaTodosOsVereadores.end(); vereador++){
            Vereador *auxVereador = *vereador;
            if (vereadorMaisVotado->getVotos()<=auxVereador->getVotos()){
                it = vereador;
                vereadorMaisVotado = auxVereador;								
                listaIncompleta=true;								
            }															
        }
        if (listaIncompleta){
            votosTotais += vereadorMaisVotado->getVotos();
            vereadoresOrdenados.push_back(vereadorMaisVotado);
            it = listaTodosOsVereadores.erase(it);
            vereadorMaisVotado = new Vereador();
        }														
    }
    listaTodosOsVereadores = vereadoresOrdenados;
}

void EleicaoDeVereadores::ordenaPartidosPorVotos(){
    Partido *partidoMaisVotado = new Partido();
    list<Partido*> partidosOrdenados;
    list<Partido*>::iterator it;
    bool listaIncompleta = true;
    while (listaIncompleta){											
        listaIncompleta = false;									
        for (list<Partido*>::iterator partido = listaTodosOsPartidos.begin(); partido != listaTodosOsPartidos.end(); partido++){
            Partido *auxPartido = *partido;
            if (partidoMaisVotado->getVotosPartido()<=auxPartido->getVotosPartido()){
                it = partido;
                partidoMaisVotado = auxPartido;								
                listaIncompleta=true;									
            }														
        }
        if (listaIncompleta){
            partidosOrdenados.push_back(partidoMaisVotado);
            it = listaTodosOsPartidos.erase(it);
            partidoMaisVotado = new Partido();
        }														
    }
    listaTodosOsPartidos = partidosOrdenados;
}

void EleicaoDeVereadores::ordenaColigacoesPorVotos(){
    Coligacao *coligacaoMaisVotada = new Coligacao();
    list<Coligacao*> coligacoesOrdenadas;
    list<Coligacao*>::iterator it;
    bool listaIncompleta = true;
    while (listaIncompleta){											
        listaIncompleta = false;									
        for (list<Coligacao*>::iterator coligacao = listaTodasAsColigacoes.begin(); coligacao != listaTodasAsColigacoes.end(); coligacao++){
            Coligacao *auxColigacao = *coligacao;
            if (coligacaoMaisVotada->getVotosColigacao()<=auxColigacao->getVotosColigacao()){
                it = coligacao;
                coligacaoMaisVotada = auxColigacao;								
                listaIncompleta=true;										
            }														
        }
        if (listaIncompleta){
            coligacoesOrdenadas.push_back(coligacaoMaisVotada);
            it = listaTodasAsColigacoes.erase(it);
            coligacaoMaisVotada = new Coligacao();
        }														
    }
    listaTodasAsColigacoes = coligacoesOrdenadas;
}

void EleicaoDeVereadores::vereadoresEleitos(){
    cout<<"Vereadores eleitos:"<<endl;
    for (list<Vereador*>::iterator vereador = listaTodosOsVereadores.begin(); vereador != listaTodosOsVereadores.end(); vereador++){
        Vereador *auxVereador = *vereador;
        if (auxVereador->isEleito()){
            if (auxVereador->getColigacao()->getNome().compare(auxVereador->getPartido()->getNome())==0){
                cout<<auxVereador->getId()<<" - "<<auxVereador->getNome()<<" ("<<auxVereador->getPartido()->getNome()<<", "<<auxVereador->getVotos()<<" votos)"<<endl;	
            }
            else{
                cout<<auxVereador->getId()<<" - "<<auxVereador->getNome()<<" ("<<auxVereador->getPartido()->getNome()<<", "<<auxVereador->getVotos()<<" votos) - Coligação: "<<auxVereador->getColigacao()->getNome()<<endl;
            }
        }
    }
    cout<<endl;
}

void EleicaoDeVereadores::vereadoresMaisVotados(){
    int i=0;
    cout<<"Candidatos mais votados (em ordem decrescente de votação e respeitando número de vagas):"<<endl;
    for (list<Vereador*>::iterator vereador = listaTodosOsVereadores.begin(); vereador != listaTodosOsVereadores.end(); vereador++){
        Vereador *auxVereador = *vereador;
        if (i<eleitos){
            i+=1;
            if (auxVereador->getColigacao()->getNome().compare(auxVereador->getPartido()->getNome())==0){
                cout<<i<<" - "<<auxVereador->getNome()<<" ("<<auxVereador->getPartido()->getNome()<<", "<<auxVereador->getVotos()<<" votos)"<<endl;	
            }														
            else{
                cout<<i<<" - "<<auxVereador->getNome()<<" ("<<auxVereador->getPartido()->getNome()<<", "<<auxVereador->getVotos()<<" votos) - Coligação: "<<auxVereador->getColigacao()->getNome()<<endl;
            }
        }
    }
    cout<<endl;
}

void EleicaoDeVereadores::seriamEleitosSeMajoritario(){
    int i=0;
    cout<<"Teriam sido eleitos se a votação fosse majoritária, e não foram eleitos:"<<endl<<"(com sua posição no ranking de mais votados)"<<endl;
    for (list<Vereador*>::iterator vereador = listaTodosOsVereadores.begin(); vereador != listaTodosOsVereadores.end(); vereador++){
        Vereador *auxVereador = *vereador;
        if (i<eleitos){
            i+=1;
            if (!(auxVereador->isEleito())){
                if (auxVereador->getColigacao()->getNome().compare(auxVereador->getPartido()->getNome())==0){
                    cout<<i<<" - "<<auxVereador->getNome()<<" ("<<auxVereador->getPartido()->getNome()<<", "<<auxVereador->getVotos()<<" votos)"<<endl;	
                }														
                else{
                    cout<<i<<" - "<<auxVereador->getNome()<<" ("<<auxVereador->getPartido()->getNome()<<", "<<auxVereador->getVotos()<<" votos) - Coligação: "<<auxVereador->getColigacao()->getNome()<<endl;
                }
            }
        }
    }
    cout<<endl;
}

void EleicaoDeVereadores::foramEleitosMinoritario(){
    int i=0;
    cout<<"Eleitos, que se beneficiaram do sistema proporcional:"<<endl<<"(com sua posição no ranking de mais votados)"<<endl;
    for (list<Vereador*>::iterator vereador = listaTodosOsVereadores.begin(); vereador != listaTodosOsVereadores.end(); vereador++){
        Vereador *auxVereador = *vereador;
        i+=1;
        if (i>eleitos){      
            if (auxVereador->isEleito()){
                if (auxVereador->getColigacao()->getNome().compare(auxVereador->getPartido()->getNome())==0){
                    cout<<i<<" - "<<auxVereador->getNome()<<" ("<<auxVereador->getPartido()->getNome()<<", "<<auxVereador->getVotos()<<" votos)"<<endl;	
                }														
                else{
                    cout<<i<<" - "<<auxVereador->getNome()<<" ("<<auxVereador->getPartido()->getNome()<<", "<<auxVereador->getVotos()<<" votos) - Coligação: "<<auxVereador->getColigacao()->getNome()<<endl;
                }
            }
        }
    }
    cout<<endl;
}

void EleicaoDeVereadores::imprimeListaColigacoes(){
    int i=0;
    cout<<"Votação (nominal) das coligações e número de candidatos eleitos:"<<endl;
    for (list<Coligacao*>::iterator coligacao = listaTodasAsColigacoes.begin(); coligacao != listaTodasAsColigacoes.end(); coligacao++){
        Coligacao* auxCol = *coligacao;
        i+=1;
        if(auxCol->getVereadoresEleitos()<=1){
            cout<<i<<" - Coligação: "<<auxCol->getNome()<<", "<<auxCol->getVotosColigacao()<<" votos, "<<auxCol->getVereadoresEleitos()<<" canditado eleito"<<endl;		
        }																				
        else{
            cout<<i<<" - Coligação: "<<auxCol->getNome()<<", "<<auxCol->getVotosColigacao()<<" votos, "<<auxCol->getVereadoresEleitos()<<" canditados eleitos"<<endl;
        }
    }
    cout<<endl;
}

void EleicaoDeVereadores::imprimeListaPartidos(){
    cout<<"Votação (nominal) dos partidos e número de candidatos eleitos:"<<endl;
    int i=0;
    for (list<Partido*>::iterator partido = listaTodosOsPartidos.begin(); partido != listaTodosOsPartidos.end(); partido++){
        Partido* auxPar = *partido;
        i+=1;
        if (auxPar->getVereadoresEleitos()<=1){
            cout<<i<<" - "<<auxPar->getNome()<<", "<<auxPar->getVotosPartido()<<" votos, "<<auxPar->getVereadoresEleitos()<<" canditado eleito"<<endl;			
        }
        else{
            cout<<i<<" - "<<auxPar->getNome()<<", "<<auxPar->getVotosPartido()<<" votos, "<<auxPar->getVereadoresEleitos()<<" canditados eleitos"<<endl;
        }
    }
    cout<<endl<<"Total de votos nominais: "<<votosTotais<<endl;
}

EleicaoDeVereadores::EleicaoDeVereadores() {
}

EleicaoDeVereadores::EleicaoDeVereadores(const EleicaoDeVereadores& orig) {
}

EleicaoDeVereadores::~EleicaoDeVereadores() {
}

